rustls-symcrypt/src/signer.rs

```diff
@@ -1,22 +1,321 @@
+
+        match self.scheme {
+            SignatureScheme::RSA_PKCS1_SHA256 => {
+                let hashed_message = sha256(message);
+                match self.key.pkcs1_sign(&hashed_message, HashAlgorithm::Sha256) {
+                    Ok(signature) => Ok(signature),
+                    Err(e) => Err(Error::General(format!("failed to sign message: {}", e))),
+                }
+            }
+            SignatureScheme::RSA_PKCS1_SHA384 => {
+                let hashed_message = sha384(message);
+                match self.key.pkcs1_sign(&hashed_message, HashAlgorithm::Sha384) {
+                    Ok(signature) => Ok(signature),
+                    Err(e) => Err(Error::General(format!("failed to sign message: {}", e))),
+                }
+            }
+            SignatureScheme::RSA_PKCS1_SHA512 => {
+                let hashed_message = sha512(message);
+                match self.key.pkcs1_sign(&hashed_message, HashAlgorithm::Sha512) {
+                    Ok(signature) => Ok(signature),
+                    Err(e) => Err(Error::General(format!("failed to sign message: {}", e))),
+                }
+            }
+            SignatureScheme::RSA_PSS_SHA256 => {
+                let hashed_message = sha256(message);
+                match self.key.pss_sign(&hashed_message, HashAlgorithm::Sha256, 32) {
+                    Ok(signature) => Ok(signature),
+                    Err(e) => Err(Error::General(format!("failed to sign message: {}", e))),
+                }
+            }
+            SignatureScheme::RSA_PSS_SHA384 => {
+                let hashed_message = sha384(message);
+                match self.key.pss_sign(&hashed_message, HashAlgorithm::Sha384, 48) {
+                    Ok(signature) => Ok(signature),
+                    Err(e) => Err(Error::General(format!("failed to sign message: {}", e))),
+                }
+            }
+            SignatureScheme::RSA_PSS_SHA512 => {
+                let hashed_message = sha512(message);
+                match self.key.pss_sign(&hashed_message, HashAlgorithm::Sha512, 64) {
+                    Ok(signature) => Ok(signature),
+                    Err(e) => Err(Error::General(format!("failed to sign message: {}", e))),
+                }
+            }
+            _ => {
+                return Err(Error::General("unsupported RSA-PKCS1 signature scheme".into()));
+            }
+        }
+    }
+
+    fn scheme(&self) -> SignatureScheme {
+        self.scheme
+    }
+}
+
+static ALL_ECDSA_SCHEMES: &[SignatureScheme] = &[
+    SignatureScheme::ECDSA_NISTP256_SHA256,
+    SignatureScheme::ECDSA_NISTP384_SHA384,
+    SignatureScheme::ECDSA_NISTP521_SHA512,
+];
+
+#[derive(Debug)]
+pub struct EcdsaSigningKey {
+    key: Arc<EcKey>, 
+}
+
+
+impl EcdsaSigningKey {
+    pub fn new(der: &PrivateKeyDer<'_>) -> Result<Self, Error> {
+        let key = match der {
+            PrivateKeyDer::Pkcs1(pkcs1) => {
```

Comment:
>:: [](http://example.com/codeflow?start=25&length=2)

[https://docs.rs/sec1/latest/sec1/struct.EcPrivateKey.html](https://docs.rs/sec1/latest/sec1/struct.EcPrivateKey.html)

```
ECPrivateKey ::= SEQUENCE {
  version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
  privateKey     OCTET STRING,
  parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
  publicKey  [1] BIT STRING OPTIONAL
}
```
the params will have the curve type

map this to rust-symcrypt struct EcCurve

[https://github.com/microsoft/rust-symcrypt/blob/9c30f5d8dc05a5db4156ee685ed1dfd33f8f53a3/rust-symcrypt/src/ecc/mod.rs#L72](https://github.com/microsoft/rust-symcrypt/blob/9c30f5d8dc05a5db4156ee685ed1dfd33f8f53a3/rust-symcrypt/src/ecc/mod.rs#L72)